### PR TITLE
feat: Allow to add annotations when che server create namspace like labels

### PR DESCRIPTION
**Pull request by disaster37**
Signed-off-by: disaster37 <linuxworkgroup@hotmail.com>

### What does this PR do?

It allow to actively annotate namespace like is the case for labels.

My use case is to use eclpise CHE on top of rancher2. Namepsace must be part of `project` (rancher concept) to have the right Pod Security Policy and to look namepace on UI/cli.

To do that, che server must be add the following labels/annotations when it create namespace for users:
  - label: `field.cattle.io/projectId=p-8j45h`
  - annotation: `field.cattle.io/projectId=c-57d7n:p-8j45h`

Like we exchange with @skabashnyuk on chat, it can be set on operator with this options:
```
customCheProperties:
      CHE_INFRA_KUBERNETES_NAMESPACE_ANNOTATIONS: che.eclipse.org/username=<username>,field.cattle.io/projectId=c-57d7n:p-8j45h
      CHE_INFRA_KUBERNETES_NAMESPACE_LABELS: app.kubernetes.io/part-of=che.eclipse.org,app.kubernetes.io/component=workspaces-namespace,field.cattle.io/projectId=p-8j45h
```

### What issues does this PR fix or reference?

Maybee fix #16426

### How to test this PR?

Deploy che server with chectl and use operator.
On CR cheCluster, set:

```
customCheProperties:
      CHE_INFRA_KUBERNETES_NAMESPACE_ANNOTATIONS: che.eclipse.org/username=<username>,field.cattle.io/projectId=c-57d7n:p-8j45h
      CHE_INFRA_KUBERNETES_NAMESPACE_LABELS: app.kubernetes.io/part-of=che.eclipse.org,app.kubernetes.io/component=workspaces-namespace,field.cattle.io/projectId=p-8j45h
```

> I haven't found a way to build che server docker image on local to test it ? 

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [X ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [X ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [X ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [X ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 1]
```diff
@@ -309,6 +309,9 @@ che.infra.kubernetes.namespace.default=<username>-che
 # Defines whether che-server should try to label the workspace namespaces.
 che.infra.kubernetes.namespace.label=true
 
+# Defines whether che-server should try to annotate the workspace namespaces.
+che.infra.kubernetes.namespace.annotate=true
```
Comment (skabashnyuk) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:313]: why do we need this flag? why don't we act as we do with labels?

File: devfile.yaml [hunk 2]
```diff
@@ -18,4 +18,4 @@ commands:
       - type: exec
         component: maven
         command: mvn clean install -DskipTests
-        workdir: ${CHE_PROJECTS_ROOT}/che
+        workdir: ${CHE_PROJECTS_ROOT}/che-server
```
Comment (skabashnyuk) [inline devfile.yaml:21]: can we make non-subject-related changes in separate pr?

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 1 above)
Comment (disaster37) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:313]: Maybee somebody should only create label or annotation, but not the two.
If we should only one flag to drive or not label and annotation, maybee the flag `che.infra.kubernetes.namespace.label` must be renamed ?

Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:313]: I would expect some kind of complementary behavior. don't you agree?

Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:313]: Let's rename it to `che.infra.kubernetes.namespace.annotation` to be consistent with labels.

**Comment by skabashnyuk** [comment]
Tested on OpenShift.

Знімок екрана 2021-07-21 о 10 42 43
Looks like placeholders are not provisioned.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 1 above)
Comment (disaster37) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: it's done.

File: devfile.yaml (hunk 2 above)
Comment (disaster37) [reply devfile.yaml]: I  have removed this quick fix.

**Comment by skabashnyuk** [comment]
@disaster37 could you please take a look at placeholders provisioning?

**Comment by disaster37** [comment]
Hi, I doesn't understand what you need ?

I have tested with your image `quay.io/skabashn/che-server:BRANCHNAME57` on rancher environment. It's work like a charm.
But I think in `che.eclipse.org/username: <username>` that `<username>` must replace by real user name ?

**Comment by skabashnyuk** [comment]
>But I think in che.eclipse.org/username: <username> that <username> must replace by real user name ?

yes. 
this method  https://github.com/eclipse-che/che-server/blob/0860edadd32aac261803e001787a5d2d2da28489/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java#L576-L586 has to be adjusted.

**Comment by disaster37** [comment]
I doesn't found the way to implements test that valid it.

I check somethink like this, but it doesn't work...

```java
@Test
  public void testUsernamePlaceholderInAnnotationsIsEvaluated() throws InfrastructureException {
    List<Namespace> namespaces =
        singletonList(
            new NamespaceBuilder()
                .withNewMetadata()
                .withName("jondoe-che")
                .withAnnotations(Map.of(NAMESPACE_ANNOTATION_NAME, "jondoe"))
                .endMetadata()
                .withNewStatus()
                .withNewPhase("Active")
                .endStatus()
                .build());
    doReturn(namespaces).when(namespaceList).getItems();

    namespaceFactory =
        new KubernetesNamespaceFactory(
            "",
            "",
            "<username>-che",
            true,
            true,
            true,
            NAMESPACE_LABELS,
            "try_placeholder_here=<username>",
            clientFactory,
            cheClientFactory,
            userManager,
            preferenceManager,
            pool);
    EnvironmentContext.getCurrent().setSubject(new SubjectImpl("jondoe", "123", null, false));
    namespaceFactory.list();

    //@TODO
   // How I can check that the placeholder `<username>` on annotation as been replaced by value jondoe ? 
    
  }
```

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java [hunk 3]
```diff
@@ -204,6 +211,58 @@ protected void label(Namespace namespace, Map<String, String> ensureLabels)
...
+                  .build());
+    } catch (KubernetesClientException kce) {
+      if (kce.getCode() == 403) {
+        LOG.warn(
+            "Can't annotate the namespace due to lack of permissions. Grant cluster-wide permissions "
+                + "to `get` and `update` the `namespaces` to the `che` service account "
+                + "(Che operator might have already prepared a cluster role called "
+                + "`che-namespace-editor` for this, depending on its configuration). "
+                + "Alternatively, consider disabling the feature by setting "
+                + "`che.infra.kubernetes.namepsace.annotation` to `false`.");
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java:259]: Copy pasta including the typo :)

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 4]
```diff
@@ -309,6 +309,9 @@ che.infra.kubernetes.namespace.default=<username>-che
 # Defines whether che-server should try to label the workspace namespaces.
 che.infra.kubernetes.namespace.label=true
 
+# Defines whether che-server should try to annotate the workspace namespaces.
+che.infra.kubernetes.namespace.annotation=true
```
Comment (metlos) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:313]: I would keep the `annotate` name here... It makes sense for it to be a verb in infinitive as opposed to a noun being set to `true` or `false`.

It also makes sense together with `...annotations`, e.g. `annotate` (true or false) with these `annotations`.

**Comment by metlos** [comment]
> I doesn't found the way to implements test that valid it.
> 
> I check somethink like this, but it doesn't work...
> 
> ```java
> @Test
>   public void testUsernamePlaceholderInAnnotationsIsEvaluated() throws InfrastructureException {
>     List<Namespace> namespaces =
>         singletonList(
>             new NamespaceBuilder()
>                 .withNewMetadata()
>                 .withName("jondoe-che")
>                 .withAnnotations(Map.of(NAMESPACE_ANNOTATION_NAME, "jondoe"))
>                 .endMetadata()
>                 .withNewStatus()
>                 .withNewPhase("Active")
>                 .endStatus()
>                 .build());
>     doReturn(namespaces).when(namespaceList).getItems();
> 
>     namespaceFactory =
>         new KubernetesNamespaceFactory(
>             "",
>             "",
>             "<username>-che",
>             true,
>             true,
>             true,
>             NAMESPACE_LABELS,
>             "try_placeholder_here=<username>",
>             clientFactory,
>             cheClientFactory,
>             userManager,
>             preferenceManager,
>             pool);
>     EnvironmentContext.getCurrent().setSubject(new SubjectImpl("jondoe", "123", null, false));
>     namespaceFactory.list();
> 
>     //@TODO
>    // How I can check that the placeholder `<username>` on annotation as been replaced by value jondoe ? 
>     
>   }
> ```

You can supply a mocked `cheClientFactory` to the `KubernetesNamespaceFactory` the same way as you do in your other tests and check that the namespace is updated with the appropriate annotations.

Also, please don't forget to enclose the entire test in a try-finally and reset the current subject to its original value after the test.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 4 above)
Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:313]: ok

**Comment by disaster37** [comment]
@skabashnyuk are you OK to keep property named as `che.infra.kubernetes.namespace.annotate` instead of `che.infra.kubernetes.namespace.annotation` like say @metlos ?

**Comment by skabashnyuk** [comment]
> @skabashnyuk are you OK to keep property named as che.infra.kubernetes.namespace.annotate instead of che.infra.kubernetes.namespace.annotation like say @metlos ?

Yes.
His argument's "kung fu" is better than mine.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 4 above)
Comment (disaster37) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: It's renamed to `che.infra.kubernetes.namespace.annotate`

**Comment by disaster37** [comment]
I think it's now OK.

**Comment by skabashnyuk** [comment]
[crw-ci-test --rebuild]

**Comment by skabashnyuk** [comment]
@disaster37  can you please check   https://github.com/probot/semantic-pull-requests

**Comment by disaster37** [comment]
I doesn't understand why build task failed.
On my environment, `mvn clean install -rf :infrastructure-openshift` work fine...

**Comment by skabashnyuk** [comment]
Not working for OpenShift 4.7 
Знімок екрана 2021-07-27 о 10 18 13
Знімок екрана 2021-07-27 о 10 17 58
Not sure that it's even possible. Need to check

**Comment by skabashnyuk** [comment]
Tested the wrong image. Correct ```quay.io/skabashn/che-server:namespace-annotations```
I can see these annotations
Знімок екрана 2021-07-27 о 10 59 37
Знімок екрана 2021-07-27 о 10 59 28
However, placeholders aren't evaluated.
@disaster37 can you check that in tests.
https://github.com/eclipse-che/che-server/pull/57/checks?check_run_id=3158607949 is saying that some stubbing is unnecesery.  Ideally, it has to be checked. As a temporary workaround, you can add ```org.mockito.Mockito.lenient```

**Comment by disaster37** [comment]
I foget to replace placeholder annotation on openshift project. It's now OK.

I have also tested the image `quay.io/skabashn/che-server:namespace-annotations` on Rancher2. It's work like a charm.

The `mvn -B clean install -U -Pintegration` is OK on my workspace. SO I deosn't understand why it doesn't work on githubaction.

**Comment by skabashnyuk** [comment]
[crw-ci-test --rebuild]

**Comment by skabashnyuk** [comment]
```
2021-07-28 10:54:32,339[main]             [WARN ] [o.e.c.w.i.o.p.OpenShiftProject 261]  - Trying to get namespace 'testProject', but failed because the lack of permissions.
2021-07-28 10:54:32,340[main]             [INFO ] [.c.w.i.k.n.KubernetesNamespace 231]  - annotate={}
java.lang.NullPointerException
	at java.base/java.lang.String.replace(String.java:2143)
	at org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespaceFactory.evaluateAnnotationPlaceholders(KubernetesNamespaceFactory.java:596)
	at org.eclipse.che.workspace.infrastructure.openshift.project.OpenShiftProjectFactory.getOrCreate(OpenShiftProjectFactory.java:115)
	at org.eclipse.che.workspace.infrastructure.openshift.project.OpenShiftProjectFactoryTest.shouldCallStopWorkspaceRoleProvisionWhenIdentityProviderIsDefined(OpenShiftProjectFactoryTest.java:626)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:133)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:584)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:172)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:804)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:145)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.testng.TestRunner.privateRun(TestRunner.java:770)
	at org.testng.TestRunner.run(TestRunner.java:591)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:402)
```
And the reason somewhere around this

```
    EnvironmentContext.getCurrent()
        .setSubject(new SubjectImpl(USER_NAME, USER_ID, "t-354t53xff34234", false));
  }

  @AfterMethod
  public void cleanup() {
    EnvironmentContext.reset();
  }

```
This part is setting user's date in context for placeholders.
@disaster37  can you take a look?

**Comment by disaster37** [comment]
I have no idea why the test `shouldCallStopWorkspaceRoleProvisionWhenIdentityProviderIsDefined()` (line 588) failed`. Like you say, the EnvironmentContext is right setted by `setup()` method. So there are no reason why `namespaceCtx.getUserName()` return null on `evaluateAnnotationPlaceholders` method.

More over, I can't reproduce it on my che workspace (created with devfile.yaml from this repository). When I run `mvn -B clean install -U -Pintegration` all tests are Ok. So why it doesn't work in githubactions ? It lauch the same cmd...

**Comment by skabashnyuk** [comment]
>I have no idea why the test shouldCallStopWorkspaceRoleProvisionWhenIdentityProviderIsDefined() (line 588) failed. Like you say, the EnvironmentContext is right setted by setup() method. So there are no reason why namespaceCtx.getUserName() return null on evaluateAnnotationPlaceholders method.


The reason is that  in this line of setup in OpenShiftProjectFactoryTest
```
    EnvironmentContext.getCurrent()
        .setSubject(new SubjectImpl(USER_NAME, USER_ID, "t-354t53xff34234", false));
```
method EnvironmentContext.getCurrent() return mock from KeycloakProviderConfigFactoryTest test . Calling setSubject on it basically do nothing.

**Comment by disaster37** [comment]
OK, I now understand what the hell.

**Comment by skabashnyuk** [comment]
[crw-ci-test --rebuild]

**Comment by skabashnyuk** [comment]
@disaster37 I've experimented a bit with new tests. these stubs in new tests aren't used. Could you just remove them?
Example OpenShiftProjectTest.java
```
//    KubernetesClient cheKubeClient = mock(KubernetesClient.class);
//    lenient().doReturn(cheKubeClient).when(cheClientFactory).create();
//
//    NonNamespaceOperation nonNamespaceOperation = mock(NonNamespaceOperation.class);
//    lenient().doReturn(nonNamespaceOperation).when(cheKubeClient).namespaces();
//
//    lenient()
//        .doAnswer(a -> a.getArgument(0))
//        .when(nonNamespaceOperation)
//        .createOrReplace(any(Namespace.class));

```

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceTest.java [hunk 5]
```diff
@@ -378,14 +378,112 @@ public void testDontTryToLabelNamespaceIfNoNamespacesProvided() throws Infrastru
...
+    Map<String, String> annotations =
+        Map.of("annotation.with.this", "yes", "and.this", "of courese");
+
+    Namespace namespace = prepareNamespace(NAMESPACE);
+    namespace.getMetadata().setAnnotations(annotations);
+    KubernetesNamespace kubernetesNamespace =
+        new KubernetesNamespace(clientFactory, cheClientFactory, executor, NAMESPACE, WORKSPACE_ID);
+
+    KubernetesClient cheKubeClient = mock(KubernetesClient.class);
+    lenient().doReturn(cheKubeClient).when(cheClientFactory).create();
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceTest.java:434]: lenient  stubs can be removed here and below

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceTest.java [hunk 6]
```diff
@@ -378,14 +378,112 @@ public void testDontTryToLabelNamespaceIfNoNamespacesProvided() throws Infrastru
...
+      throws InfrastructureException {
+    // given
+    Map<String, String> existingAnnotations = Map.of("some", "annotations");
+    Namespace namespace = prepareNamespace(NAMESPACE);
+    namespace.getMetadata().setAnnotations(existingAnnotations);
+    KubernetesNamespace kubernetesNamespace =
+        new KubernetesNamespace(clientFactory, cheClientFactory, executor, NAMESPACE, WORKSPACE_ID);
+
+    KubernetesClient cheKubeClient = mock(KubernetesClient.class);
+    lenient().doReturn(cheKubeClient).when(cheClientFactory).create();
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceTest.java:464]: lenient  stubs can be removed here and below

File: infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProjectTest.java [hunk 7]
```diff
@@ -445,13 +445,93 @@ public void testDontTryToLabelNamespaceIfAlreadyLabeled() throws InfrastructureE
...
+        new OpenShiftProject(
+            clientFactory,
+            cheClientFactory,
+            cheServerOpenshiftClientFactory,
+            executor,
+            PROJECT_NAME,
+            WORKSPACE_ID);
+
+    KubernetesClient cheKubeClient = mock(KubernetesClient.class);
+    lenient().doReturn(cheKubeClient).when(cheClientFactory).create();
```
Comment (skabashnyuk) [inline infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProjectTest.java:516]: lenient  stubs can be removed here and below

**Comment by disaster37** [comment]
I have cleaned the test as you asked.

**Comment by skabashnyuk** [comment]
[crw-ci-test --rebuild]

**Comment by disaster37** [comment]
Have you any idea why test failed ? It seems not direct impact from my change ?

**Comment by skabashnyuk** [comment]
Can you please merge the latest "main" to your branch.  There is a possibility that https://github.com/eclipse-che/che-server/pull/69 can help.

**Comment by skabashnyuk** [comment]
Tested on mikube
Знімок екрана 2021-07-30 о 13 37 37



Tested on OpenShift 4.7

Знімок екрана 2021-07-30 о 13 45 45

**Review by skabashnyuk — APPROVED** [review]

**Comment by disaster37** [comment]
Hi, It's merge (upstream/main)

**Comment by skabashnyuk** [comment]
[crw-ci-test --rebuild]

**Review by mshaposhnik — APPROVED** [review]

**Review by sparkoo — APPROVED** [review]
lgtm, thanks

**Pull request merged**